src/Bicep.Core/Text/SpellChecker.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Bicep.Core.Text
+{
+    public static class SpellChecker
+    {
+        public static string? GetSpellingSuggestion(string name, IEnumerable<string> candidates)
+        {
+            if (candidates?.Any() != true)
+            {
+                return null;
+            }
+
+            // 0.34 allows 1 insertion/deletion/substitution for every 3 characters.
+            int maxLengthDifference = Min((int)Math.Floor(name.Length * 0.34), 2);
+            int maxDistance = (int)Math.Floor(name.Length * 0.34) + 1;
+
+            string? bestCandidate = null;
+            bool checkCaseInsensitiveEqualityOnly = false;
+            string nameUpperCase = name.ToUpperInvariant();
+
+            foreach (string candidate in candidates)
+            {
+                if (string.IsNullOrEmpty(candidate) || Math.Abs(name.Length - candidate.Length) > maxLengthDifference)
+                {
+                    continue;
+                }
+
+                string candidateUpperCase = candidate.ToUpperInvariant();
+
+                if (nameUpperCase.Equals(candidateUpperCase))
+                {
+                    return candidate;
+                }
+
+                if (checkCaseInsensitiveEqualityOnly)
+                {
+                    continue;
+                }
+
+                if (candidate.Length < 3)

```

Comment:
The 3 here is related to 0.34 above?

Reply:
Sort of. I took this from TypeScript's implementation, their point is that user would have noticed the difference because the candidate's name is short. Initially, I didn't included the check as I was not quite convinced, plus C# doesn't seem to do that, but later one of my unit tests captured a failure where it won't produce suggestion for any name whose length is less than 3, and the reason is that 0.34 will emit `maxDistance=0` for such short names, meaning the returns value of `getLevenshteinWithMax` will always be -1. The length check on candidates happened to eliminate that edge case, so I added it back, otherwise we'll have to conditionally set `maxDistance` based on the length of name.